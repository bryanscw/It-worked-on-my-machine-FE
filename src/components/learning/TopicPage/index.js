import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';

import Loader from '../../common/Loader';
import { retrieveTopic, selectTopic, selectTopicLoading, selectTopicFailed } from '../../../redux/ducks/topics';
import { listLevels, selectLevelsLoading, selectLevelsFailed, selectPlayableLevels } from '../../../redux/ducks/levels';

class TopicPage extends Component {
    componentDidMount() {
        const topicId = parseInt(this.props.match.params.topicId);
        this.props.retrieveTopic(topicId);
        this.props.listLevels(topicId);
    }

    render() {
        const {
            topicLoading,
            topicFailed,
            topic,
            levelsLoading,
            levelsFailed,
            levels,
        } = this.props;
        
        if (topicLoading || levelsLoading)
            return <Loader />;

        if (topicFailed || !topic)
            return <Redirect to="/not-found" />;
        
        return (
            <div className="container">
                <Link className="btn btn-light mb-2" to="/">
                    <FontAwesomeIcon icon={faChevronLeft}/> Back to Home
                </Link>
                <h1>{topic.title}</h1>
                <p>{topic.description}</p>
                <h2>Levels</h2>
                {
                    levels.length !== 0 && !levelsFailed
                        ? levels.map((level) => (
                            <div href="#" className="card mb-4" key={level.id}>
                                <div className="card-body">
                                    <Link to={`/topics/${topic.id}/levels/${level.id}`}>
                                        <h3 className="card-title">{level.title}</h3>
                                    </Link>
                                    <p className="card-text">{level.description}</p>
                                </div>
                            </div>
                        ))
                        : <p>No levels found.</p>
                }
            </div>
        );
    }
}

TopicPage.propTypes = {
    match: PropTypes.object.isRequired,

    topicLoading: PropTypes.bool.isRequired,
    topicFailed: PropTypes.bool.isRequired,
    topic: PropTypes.object,
    levelsLoading: PropTypes.bool.isRequired,
    levelsFailed: PropTypes.bool.isRequired,
    levels: PropTypes.array.isRequired,

    retrieveTopic: PropTypes.func.isRequired,
    listLevels: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
    topicLoading: selectTopicLoading(state),
    topicFailed: selectTopicFailed(state),
    topic: selectTopic(state),
    levelsLoading: selectLevelsLoading(state),
    levelsFailed: selectLevelsFailed(state),
    levels: selectPlayableLevels(state),
});

const dispatchers = {
    retrieveTopic,
    listLevels
};

export default connect(mapStateToProps, dispatchers)(TopicPage);