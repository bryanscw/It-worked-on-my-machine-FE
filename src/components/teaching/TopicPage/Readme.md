```js 
import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap/dist/js/bootstrap.js';

import { METHODS } from '../../../redux/ducks/apiHelper.js';
import { withReduxRouter } from '../../../utils/mock.js';
const matchObject = {
    params: {
        topicId: 1,
        levelId: 1,
    }
}
const initialState = {
    topicsReducer: {
        item: 
            {
                "id": 1,
                "title": "Addition",
                "description": "Putting 2 and 2 together",
            },
        
        isLoading: {
            [METHODS.LIST]: false
        },
        hasFailed: {
            [METHODS.LIST]: false
        }
    },
    levelsReducer: {
        items: [
            {
                    "id": 1,
                    "title": "Adding 1 to 100",
                    "description": "Smol numbers.",
                    "playable":false
                },
                {
                    "id": 2,
                    "title": "Adding 1 to 1000",
                    "description": "Big numbers.",
                    "playable":false
                }
        ],
        isLoading: {
            [METHODS.LIST]: false
        },
        hasFailed: {
            [METHODS.LIST]: false
        },
    }
}

withReduxRouter(<TopicPage match={matchObject} />, initialState).renderedComponent
```