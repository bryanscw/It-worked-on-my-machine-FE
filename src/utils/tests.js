import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    return <Provider store={store}>{component}</Provider>;
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    return renderWithRedux(renderWithRouter(component, route), initialState);
}