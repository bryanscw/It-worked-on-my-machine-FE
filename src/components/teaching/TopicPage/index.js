import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEdit, faTrash, faPlus, faChevronLeft } from '@fortawesome/free-solid-svg-icons';

import Loader from '../../common/Loader';
import ModalForm from '../../common/ModalForm';
import SimpleForm from '../SimpleForm';
import DeleteForm from '../DeleteForm';
import { retrieveTopic, selectTopic, selectTopicLoading, selectTopicFailed } from '../../../redux/ducks/topics';
import { createLevel, updateLevel, deleteLevel, listLevels, selectLevelsLoading, selectLevelsFailed, selectLevels } from '../../../redux/ducks/levels';
import { CREATE, UPDATE, DELETE, EMPTY } from '../../../utils/constants';

class TopicPage extends Component {
    constructor(props) {
        super(props);

        const topicID = parseInt(props.match.params.topicID);
        props.retrieveTopic(topicID);
        props.listLevels();

        this.state = {
            modalForm: {
                isVisible: false,
                type: null,
                selectedLevel: null,
            },
        }
    }

    openModalForm = (type, selectedLevel) => {
        this.setState({
            modalForm: {
                isVisible: true,
                type: type,
                selectedLevel: selectedLevel,
            }
        });
    }

    handleModalClose = () => {
        this.setState({
            modalForm: {
                ...this.state.modalForm,
                isVisible: false,
            }
        });
    }

    getModalFormComponent = () => {
        const {
            modalForm: {
                isVisible,
                type,
                selectedLevel
            },
        } = this.state;

        const {
            topic
        } = this.props;

        switch (type) {
            case CREATE:
                return (
                    <ModalForm
                        title="Create a Level"
                        isVisible={isVisible}
                        onClose={this.handleModalClose}
                        FormComponent={SimpleForm}
                        initialState={EMPTY}
                        onSubmit={level => this.props.createLevel({...level, topic: topic.id})}
                        />
                );

            case UPDATE:
                return (
                    <ModalForm
                        title="Edit Level"
                        isVisible={isVisible}
                        onClose={this.handleModalClose}
                        FormComponent={SimpleForm}
                        initialState={selectedLevel}
                        onSubmit={level => this.props.updateLevel({...level, id: selectedLevel.id})}
                        />
                );

            case DELETE:
                return (
                    <ModalForm
                        title="Delete Level"
                        isVisible={isVisible}
                        onClose={this.handleModalClose}
                        FormComponent={DeleteForm}
                        onSubmit={isConfirm => isConfirm && this.props.deleteLevel(selectedLevel.id)}
                        />
                );

            default:
                return null;
        }
    }

    render() {
        const {
            topicLoading,
            topicFailed,
            topic,
            levelsLoading,
            levelsFailed,
            levels,
        } = this.props;
        
        if (topicLoading || levelsLoading)
            return <Loader />;

        if (topicFailed || !topic)
            return <Redirect to="/not-found" />;

        const modalFormComponent = this.getModalFormComponent();
        
        return (
            <div className="container">
                <Link className="btn btn-light mb-2" to="/">
                    <FontAwesomeIcon icon={faChevronLeft}/> Back to Home
                </Link>
                <h1>{topic.title}</h1>
                <h2>Levels</h2>
                <div className="mb-4">
                    <button className="btn btn-primary" onClick={() => this.openModalForm(CREATE, null)}>
                        <FontAwesomeIcon icon={faPlus} className="mr-2" />Create a Level
                    </button>
                </div>
                {
                    levels.length !== 0 && !levelsFailed
                        ? levels.map((level) => (
                            <div href="#" className="card mb-4" key={level.id}>
                                <div className="card-body">
                                    <h3 className="card-title"><Link to={`/levels/${level.id}`}>{level.title}</Link> {level.isPlayable ?  <span className="badge badge-success">Playable</span> : <span className="badge badge-secondary">Unplayable</span>}</h3>
                                    <p className="card-text">{level.description}</p>
                                    <div>
                                        <button className="ml-auto btn btn-success mr-2" onClick={() => this.openModalForm(UPDATE, level)}>
                                            <FontAwesomeIcon icon={faEdit} />
                                        </button>
                                        <button className="ml-auto btn btn-danger" onClick={() => this.openModalForm(DELETE, level)}>
                                            <FontAwesomeIcon icon={faTrash} />
                                        </button>
                                    </div>
                                </div>
                            </div>
                        ))
                        : <p>No levels found.</p>
                }
                {modalFormComponent}
            </div>
        );
    }
}

TopicPage.propTypes = {
    match: PropTypes.object.isRequired,

    topicLoading: PropTypes.bool.isRequired,
    topicFailed: PropTypes.bool.isRequired,
    topic: PropTypes.object,
    levelsLoading: PropTypes.bool.isRequired,
    levelsFailed: PropTypes.bool.isRequired,
    levels: PropTypes.array.isRequired,

    retrieveTopic: PropTypes.func.isRequired,
    createLevel: PropTypes.func.isRequired,
    updateLevel: PropTypes.func.isRequired,
    deleteLevel: PropTypes.func.isRequired,
    listLevels: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => {
    const topicID = parseInt(ownProps.match.params.topicID);

    return {
        topicLoading: selectTopicLoading(state),
        topicFailed: selectTopicFailed(state),
        topic: selectTopic(state, topicID),
        levelsLoading: selectLevelsLoading(state),
        levelsFailed: selectLevelsFailed(state),
        levels: selectLevels(state, topicID),
    }
};

const dispatchers = {
    retrieveTopic,
    createLevel,
    updateLevel,
    deleteLevel,
    listLevels
};

export default connect(mapStateToProps, dispatchers)(TopicPage);