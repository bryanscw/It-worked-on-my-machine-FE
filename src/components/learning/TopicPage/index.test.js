import React from 'react'
import { render, cleanup, waitForElement } from '@testing-library/react'
import axiosMock from 'axios'

import { renderWithReduxRouter } from '../../../utils/tests.js';
import TopicPage from './index.js';

jest.mock('axios');

beforeEach(() => {
    axiosMock.get.mockResolvedValueOnce({
        data: {
            id: 1,
            title: "Addition",
            description: "Putting 2 and 2 together.",
        }
    });

    axiosMock.get.mockResolvedValueOnce({
        data: {
            content: [
                {
                    id: 1,
                    title: "Adding 1 to 100",
                    description: "Smol numbers.",
                    playable: true,
                },
                {
                    id: 2,
                    title: "Adding 1 to 1000",
                    description: "Big numbers.",
                    playable: false,
                }
            ]
        }
    });
});

afterEach(() => {
    cleanup();
    jest.clearAllMocks();
});

const matchObject = {
    params: {
        topicId: 1
    }
}

it('should take a snapshot', async () => {
    const { asFragment, getByText } = renderWithReduxRouter(<TopicPage match={matchObject} />);
    await waitForElement(() => getByText('Levels'));
    expect(asFragment()).toMatchSnapshot();
})

it('should load levels', () => {
    const { getByTestId } = renderWithReduxRouter(<TopicPage match={matchObject} />); 
    expect(getByTestId('loader')).toBeTruthy();
});

it('should load and display the levels', async () => {
    const { getByText } = renderWithReduxRouter(<TopicPage match={matchObject} />);
    const levels = await waitForElement(() => getByText('Levels'));
    expect(axiosMock.get).toHaveBeenCalledTimes(2);
    expect(levels).toBeTruthy();
})

it('should not display unplayable levels', async () => {
    const { getByText, queryByText } = renderWithReduxRouter(<TopicPage match={matchObject} />);
    const playableLevel = await waitForElement(() => getByText('Adding 1 to 100'));
    expect(playableLevel).toBeTruthy();
    const unplayableLevel = queryByText('Adding 1 to 1000');
    expect(unplayableLevel).toBeFalsy();
})

it('should not display edit or delete buttons', async () => {
    const { getByText, container } = renderWithReduxRouter(<TopicPage match={matchObject} />);
    
    await waitForElement(() => getByText('Levels'));
    
    const editButton = container.querySelector('[data-icon="edit"]');
    expect(editButton).toBeFalsy();

    const deleteButton = container.querySelector('[data-icon="trash"]');
    expect(deleteButton).toBeFalsy();
});