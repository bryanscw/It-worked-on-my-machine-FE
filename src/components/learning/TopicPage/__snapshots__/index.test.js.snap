// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <a
      class="btn btn-light mb-2"
      href="/"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left fa-w-10 "
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
        />
      </svg>
       Back to Home
    </a>
    <h1>
      Addition
    </h1>
    <p>
      Putting 2 and 2 together.
    </p>
    <h2>
      Levels
    </h2>
    <div
      class="card mb-4"
      href="#"
    >
      <div
        class="card-body"
      >
        <a
          href="/topics/1/levels/1"
        >
          <h3
            class="card-title"
          >
            Adding 1 to 100
          </h3>
        </a>
        <p
          class="card-text"
        >
          Smol numbers.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
